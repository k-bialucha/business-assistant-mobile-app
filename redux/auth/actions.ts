import {
  LOGIN,
  LOGIN_FAILURE,
  LOGIN_SUCCESS,
  LOGIN_WITH_FACEBOOK,
  LoginAction,
  LoginFailureAction,
  LoginSuccessAction,
  LoginWithFacebookAction,
  LOGOUT,
  LogoutAction,
  SET_USER_DATA,
  SetUserDataAction,
  SIGNUP,
  SIGNUP_FAILURE,
  SIGNUP_SUCCESS,
  SignupAction,
  SignupFailureAction,
  SignupSuccessAction,
} from './types';

export function login(email: string, password: string): LoginAction {
  return {
    type: LOGIN,
    payload: { email, password },
  };
}

export function loginSuccess(token: string): LoginSuccessAction {
  return {
    type: LOGIN_SUCCESS,
    payload: { token },
  };
}

export function loginFailure(message: string): LoginFailureAction {
  return {
    type: LOGIN_FAILURE,
    payload: message,
  };
}

export function logout(): LogoutAction {
  return {
    type: LOGOUT,
  };
}

export function signup(
  email: string,
  password: string,
  phone?: string
): SignupAction {
  return {
    type: SIGNUP,
    payload: { email, phone, password },
  };
}

export function signupSuccess(token: string): SignupSuccessAction {
  return {
    type: SIGNUP_SUCCESS,
    payload: { token },
  };
}

export function signupFailure(message: string): SignupFailureAction {
  return {
    type: SIGNUP_FAILURE,
    payload: message,
  };
}

export function loginWithFacebook(): LoginWithFacebookAction {
  return {
    type: LOGIN_WITH_FACEBOOK,
  };
}

export function setUserData(data): SetUserDataAction {
  return {
    type: SET_USER_DATA,
    payload: {
      name: data.name,
      id: data.id,
      image: data.image,
    },
  };
}
