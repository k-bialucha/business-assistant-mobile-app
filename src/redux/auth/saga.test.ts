import { put } from 'redux-saga/effects';

import { loginFailure, loginSuccess } from './actions';
import {
  loginSaga,
  loginWithFacebookSaga,
  loginWithGoogleSaga,
  resetPasswordSaga,
  signupSaga,
} from './saga';
import {
  DOMAIN_NAME,
  LOGIN,
  LOGIN_FAILURE,
  LOGIN_SUCCESS,
  RESET_PASSWORD,
  SIGNUP,
  SIGNUP_FAILURE,
  SIGNUP_SUCCESS,
} from './types';

jest.mock('~~env');

describe(`${DOMAIN_NAME}/saga`, () => {
  describe('loginSaga', () => {
    it('handles successful login', () => {
      const someApiResponse = {
        idToken: 'token-123',
        localId: '12345',
      };
      const mockedEmail = 'some-email';

      const generator = loginSaga({
        type: LOGIN,
        payload: { email: mockedEmail, password: 'strong-password' },
      });

      const delayDescriptor = generator.next().value;
      const apiCallDescriptor = generator.next().value;
      const putDescriptor = generator.next(someApiResponse).value;
      const saveToAsyncStorage = generator.next().value;

      expect(delayDescriptor).toMatchSnapshot();
      expect(apiCallDescriptor).toMatchSnapshot();
      expect(putDescriptor).toEqual(
        put({
          type: LOGIN_SUCCESS,
          payload: {
            token: someApiResponse.idToken,
            userData: {
              id: someApiResponse.localId,
              image: undefined,
              name: mockedEmail,
            },
          },
        })
      );
      expect(saveToAsyncStorage).toMatchSnapshot();
      expect(generator.next().done).toBe(true);
    });

    it('handles unsuccessful login', () => {
      const someError = new Error('Bad Credentials');
      const generator = loginSaga({
        type: LOGIN,
        payload: { email: 'some-email', password: 'badpass' },
      });

      const delayDescriptor = generator.next().value;
      const apiCallDescriptor = generator.next().value;
      const putDescriptor = generator.throw(someError).value;

      expect(delayDescriptor).toMatchSnapshot();
      expect(apiCallDescriptor).toMatchSnapshot();
      expect(putDescriptor).toEqual(
        put({ type: LOGIN_FAILURE, payload: someError.message })
      );
      expect(generator.next().done).toBe(true);
    });
  });

  describe('signupSaga', () => {
    it('handles successful signup', () => {
      const someApiResponse = {
        idToken: 'token-123',
        localId: '12345',
      };
      const mockedEmail = 'some-email@example.com';
      const generator = signupSaga({
        type: SIGNUP,
        payload: {
          email: mockedEmail,
          password: 'strong-password',
        },
      });

      const delayDescriptor = generator.next().value;
      const apiCallDescriptor = generator.next().value;
      const putDescriptor = generator.next(someApiResponse).value;
      const saveToAsyncStorage = generator.next().value;

      expect(delayDescriptor).toMatchSnapshot();
      expect(apiCallDescriptor).toMatchSnapshot();
      expect(putDescriptor).toEqual(
        put({
          type: SIGNUP_SUCCESS,
          payload: {
            token: someApiResponse.idToken,
            userData: {
              id: someApiResponse.localId,
              image: undefined,
              name: mockedEmail,
            },
          },
        })
      );
      expect(saveToAsyncStorage).toMatchSnapshot();
      expect(generator.next().done).toBe(true);
    });

    it('handles unsuccessful signup', () => {
      const someError = new Error('Bad Credentials');
      const generator = signupSaga({
        type: SIGNUP,
        payload: { email: 'some-email@example.com', password: 'badpass' },
      });

      const delayDescriptor = generator.next().value;
      const apiCallDescriptor = generator.next().value;
      const putDescriptor = generator.throw(someError).value;

      expect(delayDescriptor).toMatchSnapshot();
      expect(apiCallDescriptor).toMatchSnapshot();
      expect(putDescriptor).toEqual(
        put({ type: SIGNUP_FAILURE, payload: someError.message })
      );
      expect(generator.next().done).toBe(true);
    });
  });

  describe('loginWithFacebookSaga', () => {
    it('handles successful signup through facebook', () => {
      const generator: Generator = loginWithFacebookSaga();

      const facebookLoginResponse = {
        token: 'mocked-token',
        type: 'success',
        user: 'user',
      };

      const initializeFacebookApp = generator.next().value;

      expect(initializeFacebookApp).toMatchSnapshot();

      const loginWithFacebook = generator.next().value;

      expect(loginWithFacebook).toMatchSnapshot();

      // get credential to firebase auth with fb token
      const getCredentials = generator.next(facebookLoginResponse).value;

      expect(getCredentials).toMatchSnapshot();

      // sign in to firebase with received credentials
      const signToFirebaseWithCredentials = generator.next({
        token: '',
        type: '',
        user: 'user-object',
      }).value;

      expect(signToFirebaseWithCredentials).toMatchSnapshot();

      // get jwt token of created user in firebase
      const token =
        'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c';

      expect(
        generator.next({
          user: {
            getIdToken: jest.fn(() => token),
          },
        }).value
      ).toEqual(token);

      expect(generator.next(token).value).toEqual(
        put(
          loginSuccess('mocked-token', {
            name: expect.any(String),
          })
        )
      );

      expect(generator.next().value).toMatchSnapshot();

      expect(generator.next().done).toBe(true);
    });

    it('handles canceled signup through facebook', () => {
      const generator: Generator = loginWithFacebookSaga();

      generator.next();
      generator.next();

      expect(generator.next({ type: 'cancel' }).value).toEqual(
        put(loginFailure('Login to facebook canceled'))
      );

      expect(generator.next().done).toBe(true);
    });
  });

  describe('loginWithGoogleSaga', () => {
    it('handles successful signup through google', () => {
      const generator: Generator = loginWithGoogleSaga();

      const googleLoginResponse = {
        accessToken: 'mocked-token',
        idToken: 'mocked-id-token',
        type: 'success',
        user: 'user',
      };

      const loginWithGoogle = generator.next().value;

      expect(loginWithGoogle).toMatchSnapshot();

      // get credential to firebase auth with fb token
      const getCredentials = generator.next(googleLoginResponse).value;

      expect(getCredentials).toMatchSnapshot();

      // sign in to firebase with received credentials
      const signToFirebaseWithCredentials = generator.next({
        accessToken: 'mocked-token',
        idToken: 'mocked-id-token',
        type: '',
        user: 'user-object',
      }).value;

      expect(signToFirebaseWithCredentials).toMatchSnapshot();

      // get jwt token of created user in firebase
      const token =
        'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c';

      expect(
        generator.next({
          user: {
            getIdToken: jest.fn(() => token),
          },
        }).value
      ).toEqual(token);

      expect(generator.next(token).value).toEqual(
        put(
          loginSuccess(token, {
            name: expect.any(String),
          })
        )
      );

      expect(generator.next().value).toMatchSnapshot();

      expect(generator.next().done).toBe(true);
    });

    it('handles canceled signup through google', () => {
      const generator: Generator = loginWithGoogleSaga();

      generator.next();

      expect(generator.next({ type: 'cancel' }).value).toEqual(
        put(loginFailure('Login to google canceled'))
      );

      expect(generator.next().done).toBe(true);
    });
  });

  describe('resetPasswordSaga', () => {
    it('handles successful sending reset password form', () => {
      const mockedEmail = 'some@email.com';

      const generator = resetPasswordSaga({
        type: RESET_PASSWORD,
        payload: { email: mockedEmail },
      });

      const resetPasswordCall = generator.next().value;

      expect(resetPasswordCall).toMatchSnapshot();
      expect(generator.next().done).toBe(true);
    });
  });
});
