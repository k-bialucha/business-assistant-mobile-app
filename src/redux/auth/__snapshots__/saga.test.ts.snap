// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth/saga loginSaga handles successful login 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      1000,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga loginSaga handles successful login 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "some-email",
      "strong-password",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga loginSaga handles successful login 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "userData",
      "{\\"token\\":\\"token-123\\",\\"username\\":\\"some-email\\",\\"id\\":\\"12345\\"}",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga loginSaga handles unsuccessful login 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      1000,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga loginSaga handles unsuccessful login 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "some-email",
      "badpass",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga loginWithFacebookSaga handles successful signup through facebook 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "facebook-app-id",
      "Personal Business Assistant",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga loginWithFacebookSaga handles successful signup through facebook 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      Object {
        "permissions": Array [
          "public_profile",
          "email",
        ],
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga loginWithFacebookSaga handles successful signup through facebook 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "mocked-token",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga loginWithFacebookSaga handles successful signup through facebook 4`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      Object {
        "token": "",
        "type": "",
        "user": "user-object",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga loginWithFacebookSaga handles successful signup through facebook 5`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "userData",
      "{\\"token\\":\\"mocked-token\\",\\"username\\":\\"John Doe\\",\\"id\\":\\"5d4b0cbb6f82\\"}",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga loginWithGoogleSaga handles successful signup through google 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      Object {
        "androidClientId": undefined,
        "iosClientId": undefined,
        "scopes": Array [
          "profile",
          "email",
        ],
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga loginWithGoogleSaga handles successful signup through google 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "mocked-id-token",
      "mocked-token",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga loginWithGoogleSaga handles successful signup through google 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      Object {
        "accessToken": "mocked-token",
        "idToken": "mocked-id-token",
        "type": "",
        "user": "user-object",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga loginWithGoogleSaga handles successful signup through google 4`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "userData",
      "{\\"jwtToken\\":\\"eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJLYW1pbCBCaWFsdWNoYSIsImlhdCI6MTU5MDk2MzkzNCwiZXhwIjoxNjIyNDk5OTM0LCJhdWQiOiJidXNzaW5lc3MtYXNzaXN0YW50LW1vYmlsZS1hcHAiLCJzdWIiOiJqZG9lQGV4YW1wbGUuY29tIiwibmFtZSI6IkpvaG4gRG9lIiwiZW1haWwiOiJqZG9lQGV4YW1wbGUuY29tIiwidXNlcl9pZCI6IjVkNGIwY2JiNmY4MiJ9.qVrXVmcIqBeR4E6W0_8D4pq17w5NUl6s5su2U41tFIY\\",\\"username\\":\\"John Doe\\",\\"id\\":\\"5d4b0cbb6f82\\"}",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga resetPasswordSaga handles successful sending reset password form 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "some@email.com",
      Object {
        "url": "",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga signupSaga handles successful signup 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      1000,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga signupSaga handles successful signup 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "some-email@example.com",
      "strong-password",
      undefined,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga signupSaga handles successful signup 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "userData",
      "{\\"token\\":\\"token-123\\",\\"username\\":\\"some-email@example.com\\",\\"id\\":\\"12345\\"}",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga signupSaga handles unsuccessful signup 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      1000,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga signupSaga handles unsuccessful signup 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "some-email@example.com",
      "badpass",
      undefined,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
